shared/types/indexer_test.go

```
@@ -0,0 +1,120 @@
+	require.True(t, txResultsEqual(t, txResult3, txResultsFromHeight1[0]))
+	// check indexing by sender / recipient
+	sender := txResult3.GetSender()
+	recipient := txResult3.GetRecipient()
+	require.NoError(t, err)
+	txResultsFromSender, err := txIndexer.GetBySender(sender, false)
+	require.NoError(t, err)
+	txResultsFromRecipient, err := txIndexer.GetByRecipient(recipient, false)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(txResultsFromSender))
+	require.Equal(t, 1, len(txResultsFromRecipient))
+	require.True(t, txResultsEqual(t, txResult3, txResultsFromSender[0]))
+	require.True(t, txResultsEqual(t, txResult3, txResultsFromRecipient[0]))
+	// ensure it's not indexed elsewhere
+	txResultsFromSenderBad, err := txIndexer.GetBySender(recipient, false)
+	require.NoError(t, err)
+	txResultsFromRecipientBad, err := txIndexer.GetByRecipient(sender, false)
+	require.NoError(t, err)
+	require.Equal(t, 0, len(txResultsFromSenderBad))
+	require.Equal(t, 0, len(txResultsFromRecipientBad))
+}
+
+func txResultsEqual(t *testing.T, txR1, txR2 TxResult) bool {
+	bz, err := txR1.Bytes()
+	require.NoError(t, err)
+	bz2, err := txR2.Bytes()
+	require.NoError(t, err)
+	return bytes.Equal(bz, bz2)
+}
+
+// utility helpers
+
+func NewTestingTransactionResult(height, index int) TxResult {
+	testingTransaction := randLetterBytes()
+	resultCode, err := randomErr()
+	return &DefaultTxResult{
+		Tx:          testingTransaction,
+		Height:      int64(height),
+		Index:       int32(index),
+		ResultCode:  resultCode,
+		Error:       err,
+		Signer:      randomAddress(),
+		Recipient:   randomAddress(),
+		MessageType: randomMessageType(),
+	}
+}
+
+func randomMessageType() string {
+	msgTypes := []string{"send", "stake", "unstake", "editStake", "unjail"}
+	return msgTypes[rand.Intn(len(msgTypes))]
+}
+
+func randomAddress() string {
+	add, _ := crypto.GenerateAddress()
+	return add.String()
+}
+
+func randomErr() (code int32, err string) {
```

Given some of the hardcoded values, we should add a comment: E.g. `Returns an error 25% of the time`